torch/csrc/autograd/record_function.h

```diff
@@ -82,20 +118,28 @@ struct TORCH_API RecordFunction {
   Function* fn_ = nullptr;
   StringView name_;
   int64_t sequence_nr_ = -1;
-
+  std::vector<c10::IValue> inputs_;
   RecordFunction* parent_ = nullptr;
 
-  GetPackedInputsCallback inputs_cb_ = nullptr;
-  mutable bool inputs_initialized_ = false;
-  // initialized lazily by inputs_cb_
-  mutable std::vector<c10::IValue> inputs_;
+  bool initialized_ = false;
+  bool needs_inputs_ = false;
 };
 
+bool hasCallbacks();
+
+#define RECORD_FUNCTION(...) \
+  torch::autograd::profiler::RecordFunction guard; \
+  if (torch::autograd::profiler::hasCallbacks()) { \
```

Comment:
oh, I see, I think we got overcomplicated with multiple diff reviews.

I think the important point is to avoid constructing IValue here by putting explicit if (which was in the original recommendation). In order to do that we actually need to have a named argument for macro, not just varargs. I think nothing bad will happen if you just define this macro as:

```
#define RECORD_FUNCTION(name, inputs, seq_no)\
  torch::autograd::profiler::RecordFunction guard;
  if (torch::autograd::profiler::hasCallbacks()) {
    if (torch::autograd::profiler::needsInputs()) {
      guard.before(name, seq_no, inputs);
    } else {
      guard.before(name, seq_no);
    }
```
The original point was to make `input` macro argument so that it doesn't get constructed unless needed.

Yep, you won't support RECORD_FUNCTION(name, seq_no), but I'd say it's ok to throw {} there.

Reply:
I think what happens here is ok, because we check `needs_inputs_` later. So really when they are not needed, they are passed as arguments (by reference with no conversions) and left unused.

Reply:
@dzhulgakov I'm ok with using a less nice version of RECORD_FUNCTION (or even use multiple versions of macro as I already did before :) ), that would make implementation simpler

Reply:
@dzhulgakov we don't construct IValues if we don't need them, we just pass const references and that's it (as of the moment)

Reply:
@ilia-cher yeah, it's a good solution, but it might be a good idea to leave a comment nevertheless because it's not obvious.